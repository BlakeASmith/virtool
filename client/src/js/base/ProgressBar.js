import React from "react";
import styled from "styled-components";

const progressValueColor = {
    success: "#60af98",
    warning: "#be9235",
    danger: "#af3227",
    blue: "rgb(7, 104, 157)"
};

const StyledProgress = styled.progress`
    -webkit-appearance: none;
    height: 20px;
    margin-bottom: 10px;
    width: 100%;

    ::-webkit-progress-value {
        ${props => {
            return `
            background-color: ${progressValueColor[props.bsStyle]};
          `;
        }}
    }

    ::-webkit-progress-bar {
        background-color: #f5f5f5;
    }
`;

const StyledAffixedProgress = styled(StyledProgress)`
    height: 5px;
    position: absolute;

    top: 0;
    left: 0;

    ::-webkit-progress-bar {
        background-color: transparent;
    }
`;

const StyledTopProgressBar = styled(StyledProgress)`
    height: 4px;
    display: flex;
    margin-bottom: ${props => (props.marginBottom === "none" ? 0 : "10px")};

    ::-webkit-progress-bar {
        background-color: transparent;
    }
`;

export const AffixedProgressBar = ({ now, bsStyle }) => {
    return <StyledAffixedProgress max="100" value={now} bsStyle={bsStyle}></StyledAffixedProgress>;
};

export const ProgressBar = ({ now, bsStyle }) => {
    return <StyledProgress max="100" value={now} bsStyle={bsStyle}></StyledProgress>;
};

export const TopProgressBar = ({ now, bsStyle, marginBottom }) => {
    return (
        <StyledTopProgressBar
            max="100"
            value={now}
            bsStyle={bsStyle}
            marginBottom={marginBottom}
        ></StyledTopProgressBar>
    );
};
