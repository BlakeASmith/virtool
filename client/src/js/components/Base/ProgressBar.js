/**
 * @license
 * The MIT License (MIT)
 * Copyright 2015 Government of Canada
 *
 * @author
 * Ian Boyes
 *
 * @exports ProgressBar, AutoProgressBar
 *
 */

import React from "react";
import CX from "classnames";
import { round } from "lodash-es";

export class AutoProgressBar extends React.Component {

    constructor (props) {
        super(props);

        this.interval = null;
        this.state = {
            fill: 0
        };
    }

    static propTypes = {
        step: React.PropTypes.number,
        active: React.PropTypes.bool,
        interval: React.PropTypes.number,
        progressStyle: React.PropTypes.oneOf(["modal", "list"])
    };

    static defaultProps = {
        step: 0.3,
        interval: 120
    };

    componentWillReceiveProps (nextProps) {
        if (this.props.active && !nextProps.active) {
            this.stop();

            this.setState({
                fill: 1
            });
        }

        if (!this.props.active && nextProps.active) {
            this.setState({ fill: 0.1 }, () => {
                this.interval = window.setInterval(this.move, this.props.interval);
            });
        }
    }

    componentWillUnmount () {
        this.stop();
    }

    move = () => {
        // Calculate how far the bar should be filled for this iteration.
        const fill = this.state.fill + round(this.props.step * Math.random());

        // If the fill is more than 80% of the bar, stop iterating and wait for a prop change before moving the bar
        // again.
        if (fill >= 0.8) {
            this.stop();
        }

        // Update the fill so the progress bar moves.
        this.setState({
            fill: fill
        });
    };

    stop = () => function () {
        window.clearInterval(this.interval);
    };

    render () {
        return (
            <ProgressBar
                progressStyle={this.props.progressStyle}
                now={this.state.fill * 100}
            />
        );
    }

}


export default class ProgressBar extends React.PureComponent {

    constructor (props) {
        super(props);
        this.state = {};
    }

    static propTypes = {
        now: React.PropTypes.number,
        onMoved: React.PropTypes.func,
        children: React.PropTypes.element,
        progressStyle: React.PropTypes.oneOf(["modal", "list"])
    };

    static defaultProps = {
        now: 0
    };

    componentWillReceiveProps (nextProps) {
        // Listen for the bar move transition to end if the "now" prop is being changed.
        if (this.props.now !== nextProps.now) {
            this.barNode.addEventListener("transitionend", this.onTransitionend);
        }
    }

    componentWillUnmount () {
        this.barNode.removeEventListener("transitionend", this.onTransitionend);
    }

    onTransitionend = () => {
        // Stop listening for the transition end once it is complete. Only listen again if the "now" prop changes.
        this.barNode.removeEventListener("transitionend", this.onTransitionend);

        // Call the onMoved callback with the new "now" prop once the bar is done moving.
        if (this.props.onMoved) {
            this.props.onMoved(this.state.now);
        }
    };

    render () {

        const className = CX(
            "progress", {
                "progress-modal": this.props.progressStyle === "modal",
                "progress-list": this.props.progressStyle === "list"
            }
        );

        return (
            <div className={className}>
                <div
                    ref={this.barNode}
                    className="progress-bar"
                    role="progressbar"
                    ariaValuenow={this.state.now}
                    ariaValuemin="0"
                    ariaValuemax="100"
                >
                    {this.props.children}
                </div>
            </div>
        );
    }

}
