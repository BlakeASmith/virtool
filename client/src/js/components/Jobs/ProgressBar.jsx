/**
 * @license
 * The MIT License (MIT)
 * Copyright 2015 Government of Canada
 *
 * @author
 * Ian Boyes
 *
 * @exports ProgressBar
 */

'use strict';

var React = require('react');

/**
 * A component that renders a Bootstrap progress bar based on the state of a Virtool job and its progress value
 * (eg. 0.2).
 *
 * @class
 */
var ProgressBar = React.createClass({

    propTypes: {
        state: React.PropTypes.string.isRequired,
        value: React.PropTypes.number.isRequired
    },

    render: function () {
        // The label text to show in the progress bar.
        var label = '';

        // The value (0 < x < 100) used to fill the progress bar.
        var fill;

        // Classes to give the progress bar. Affect colouring and striping.
        var barClass = 'progress-bar progress-bar-';

        // A text element to be shown in an empty progress bar, indicating that the job is waiting.
        var waitingText = null;

        switch (this.props.state) {

            case 'waiting':
                // Empty bar with 'Waiting' label.
                waitingText = <span className='waiting-text'>Waiting</span>;
                fill = 0;
                break;

            case 'running':
                // Partially filled blue, animated bar with no label.
                label = null;
                barClass += 'striped active';
                fill = this.props.value * 100;
                break;

            case 'error':
                // Full red bar with 'Error' label.
                label = 'Error';
                barClass += 'danger';
                fill = 100;
                break;

            case 'cancelled':
                // Full red bar with 'Cancelled' label.
                label = 'Cancelled';
                barClass += 'danger';
                fill = 100;
                break;

            case 'complete':
                label = 'Complete';
                fill = 100;
                break;
        }

        return (
            <div className='progress'>
                {waitingText}
                <div className={barClass} aria-valuenow={fill} aria-valuemin='0' aria-valuemax='100' style={{width: fill + '%'}}>
                    <span>{label}</span>
                </div>
            </div>
        )
    }
});

module.exports = ProgressBar;