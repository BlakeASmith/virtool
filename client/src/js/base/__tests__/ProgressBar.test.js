import { ProgressBar, AffixedProgressBar } from "../ProgressBar";

describe("<ProgressBar />", () => {
    const props = {
        now: 0,
        bsStyle: ""
    };

    it("should render when [bsStyle='success']", () => {
        props.bsStyle = "success";
        const wrapper = shallow(<ProgressBar {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render when [bsStyle='warning']", () => {
        props.bsStyle = "warning";
        const wrapper = shallow(<ProgressBar {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render when [bsStyle='danger']", () => {
        props.bsStyle = "danger";
        const wrapper = shallow(<ProgressBar {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
});

describe("<AffixedProgressBar />", () => {
    const props = {
        now: 0,
        bsStyle: ""
    };

    it("should render when [bsStyle='success']", () => {
        props.bsStyle = "success";
        const wrapper = shallow(<AffixedProgressBar {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render when [bsStyle='warning']", () => {
        props.bsStyle = "warning";
        const wrapper = shallow(<AffixedProgressBar {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render when [bsStyle='danger']", () => {
        props.bsStyle = "danger";
        const wrapper = shallow(<AffixedProgressBar {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
});
