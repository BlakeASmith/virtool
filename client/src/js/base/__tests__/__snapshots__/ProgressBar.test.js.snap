// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AffixedProgressBar /> should render when [bsStyle='danger'] 1`] = `
<ProgressBar__StyledAffixedProgress
  bsStyle="danger"
  max="100"
  value={0}
/>
`;

exports[`<AffixedProgressBar /> should render when [bsStyle='success'] 1`] = `
<ProgressBar__StyledAffixedProgress
  bsStyle="success"
  max="100"
  value={0}
/>
`;

exports[`<AffixedProgressBar /> should render when [bsStyle='warning'] 1`] = `
<ProgressBar__StyledAffixedProgress
  bsStyle="warning"
  max="100"
  value={0}
/>
`;

exports[`<ProgressBar /> should render when [bsStyle='danger'] 1`] = `
<ProgressBar__StyledProgress
  bsStyle="danger"
  max="100"
  value={0}
/>
`;

exports[`<ProgressBar /> should render when [bsStyle='success'] 1`] = `
<ProgressBar__StyledProgress
  bsStyle="success"
  max="100"
  value={0}
/>
`;

exports[`<ProgressBar /> should render when [bsStyle='warning'] 1`] = `
<ProgressBar__StyledProgress
  bsStyle="warning"
  max="100"
  value={0}
/>
`;
